### PR TITLE
Implement MP3 compression as a defense against audio attacks

### PR DESCRIPTION
**Pull request by hesseltuinhof**
# Description

Add MP3 compression to the preprocessor defenses.

## Type of change

Please check all the relevant options.

- [x] New feature (non-breaking)
- [x] This change includes a documentation update

# Testing

Several tests were provided to the test suite. Furthermore, a quick demonstration of how to apply it to adversarial audio examples is added to the respective notebook.

**Test Configuration**:
- Fedora 31
- Python 3.6.10
- ART 1.3.0 development 

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/defences/preprocessor/mp3_compression.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+# SOFTWARE.
+"""
+This module implements the MP3 compression defence `Mp3Compression`.
+
+| Paper link: https://arxiv.org/abs/1801.01944
+
+| Please keep in mind the limitations of defences. For details on how to evaluate classifier security in general,
+    see https://arxiv.org/abs/1902.06705  .
+"""
+from __future__ import (absolute_import, division, print_function,
```
Comment (minhitbk) [inline art/defences/preprocessor/mp3_compression.py:26]:
```suggestion
from __future__ import absolute_import, division, print_function, unicode_literals
```

File: art/defences/preprocessor/mp3_compression.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                        unicode_literals)
```
Comment (minhitbk) [inline art/defences/preprocessor/mp3_compression.py:27]: Delete this, I think we better be consistent with this style in other files in ART.

File: art/defences/preprocessor/mp3_compression.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+from art.defences.preprocessor.preprocessor import Preprocessor
+
+logger = logging.getLogger(__name__)
+
+
+class Mp3Compression(Preprocessor):
+    """
+    Implement the MP3 compression defense approach.
+
```
Comment (minhitbk) [inline art/defences/preprocessor/mp3_compression.py:42]:
```suggestion
```

File: art/defences/preprocessor/mp3_compression.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    """
+
+    params = ["channel_index", "sample_rate"]
+
+    def __init__(self, channel_index, sample_rate, apply_fit=True, apply_predict=False):
+        """Create an instance of MP3 compression.
```
Comment (minhitbk) [inline art/defences/preprocessor/mp3_compression.py:48]:
```suggestion
        """
        Create an instance of MP3 compression.
```

File: art/defences/preprocessor/mp3_compression.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
```
Comment (minhitbk) [inline art/defences/preprocessor/mp3_compression.py:49]: Could you please give a description for sample_index and sample_rate?

File: art/defences/preprocessor/mp3_compression.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        recommended to be of type `np.int16`.
+        :type x: `np.ndarray`
+        :param y: Labels of the sample `x`. This function does not affect them in any way.
+        :type y: `np.ndarray`
+        :return: Compressed sample.
+        :rtype: `np.ndarray`
+        """
+
+        def wav_to_mp3(x, sample_rate):
+            """Apply MP3 compression to audio input of shape (samples, channel)."""
```
Comment (minhitbk) [inline art/defences/preprocessor/mp3_compression.py:83]:
```suggestion
            """
            Apply MP3 compression to audio input of shape (samples, channel).
            """
```

File: .travis.yml [hunk 7]
```diff
@@ -16,6 +17,7 @@ matrix:
         - py.test --pep8 -m pep8
 
 before_install:
+  - sudo apt-get -y -q install ffmpeg libavcodec-extra
```
Comment (minhitbk) [inline .travis.yml:20]: Should we do this in the requirements.txt file instead, where we need to describe that these libraries are essential for ART?

File: test_requirements.txt [hunk 8]
```diff
@@ -4,6 +4,7 @@ h5py>=2.10.0
 requests>=2.23.0
 statsmodels>=0.11.0
 cma>=2.7.0
+pydub>=0.23,<0.24
```
Comment (minhitbk) [inline test_requirements.txt:7]: Should we add pydub to requirements.txt?

File: tests/defences/preprocessor/test_mp3_compression.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+#
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from __future__ import (absolute_import, division, print_function,
```
Comment (minhitbk) [inline tests/defences/preprocessor/test_mp3_compression.py:18]:
```suggestion
from __future__ import absolute_import, division, print_function, unicode_literals
```

File: tests/defences/preprocessor/test_mp3_compression.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                        unicode_literals)
```
Comment (minhitbk) [inline tests/defences/preprocessor/test_mp3_compression.py:19]:
```suggestion
```

File: notebooks/adversarial_audio/model.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
+import torch.nn as nn
+
+
+class RawAudioCNN(nn.Module):
+    """Adaption of AudioNet (arXiv:1807.03418)."""
```
Comment (minhitbk) [inline notebooks/adversarial_audio/model.py:5]:
```suggestion
    """
    Adaption of AudioNet (arXiv:1807.03418).
    """
```

File: .travis.yml (hunk 7 above)
Comment (hesseltuinhof) [reply .travis.yml:20]: These are necessary system-level dependencies for `pydub`.

File: test_requirements.txt (hunk 8 above)
Comment (hesseltuinhof) [reply test_requirements.txt:7]: The `pydub` dependency is only used when someone wants to apply the `Mp3Compression` defense. It is my understanding that such less-used dependencies are the users responsibility. E.g. like the `PIL` dependency in the `JpegCompression`.

File: art/defences/preprocessor/mp3_compression.py (hunk 5 above)
Comment (hesseltuinhof) [reply art/defences/preprocessor/mp3_compression.py:49]: Great spot! Have added the missing description :+1:

**Comment by hesseltuinhof** [comment]
Thanks for your review @minhitbk. 

I agree that the introduction of the `model.py` in order to be able to run the notebook is not that elegant. Instead I now saved the model parameter dictionary and initialize the model class within the notebook. I have applied all of your comments and rebased the PR accordingly.

Please have a final look at the changes and the two remaining unresolved comments :rocket:

**Comment by beat-buesser** [comment]
@hesseltuinhof and @minhitbk Thank you very much!

**Pull request merged**